rfd/0162-machine-id-token-join-method-bot-instance.md

```diff
@@ -0,0 +1,653 @@
+be listed and deleted:
+
+```protobuf
+syntax = "proto3";
+
+package teleport.machineid.v1;
+
+service BotInstanceService {
+  // GetBotInstance returns the specified BotInstance resource.
+  rpc GetBotInstance(GetBotInstanceRequest) returns (BotInstance);
+  // ListBotInstances returns a page of BotInstance resources.
+  rpc ListBotInstances(ListBotInstancesRequest) returns (ListBotInstancesResponse);
+  // DeleteBotInstance hard deletes the specified BotInstance resource.
+  rpc DeleteBotInstance(DeleteBotInstanceRequest) returns (google.protobuf.Empty);
+}
+
+// Request for GetBotInstance.
+message GetBotInstanceRequest {
+  // The name of the bot associated with the instance.
+  string bot_name = 1;
+  // The unique identifier of the bot instance to retrieve.
+  string id = 2;
+}
+
+// Request for ListBotInstances.
+//
+// Follows the pagination semantics of
+// https://cloud.google.com/apis/design/standard_methods#list
+message ListBotInstancesRequest {
+  // The name of the Bot to list BotInstances for. If empty, all BotInstances
+  // will be listed.
+  string filter_bot_name = 1;
+  // The maximum number of items to return.
+  // The server may impose a different page size at its discretion.
+  int32 page_size = 2;
+  // The page_token value returned from a previous ListBotInstances request, if
+  // any.
+  string page_token = 3;
+}
+
+// Response for ListBotInstances.
+message ListBotInstancesResponse {
+  // BotInstance that matched the search.
+  repeated BotInstance bot_instances = 1;
+  // Token to retrieve the next page of results, or empty if there are no
+  // more results exist.
+  string next_page_token = 2;
+}
+
+// Request for DeleteBotInstance.
+message DeleteBotInstanceRequest {
+  // The name of the BotInstance to delete.
+  string bot_name = 1;
+  // The unique identifier of the bot instance to delete.
+  string id = 2;
+}
+```
+
+### Changes to the `token` Join Method
+
+As we now have a way to track the generation for a specific Bot instance, we
+can allow multiple Bot instances to be associated with a single Bot. This
+also means that the token no longer needs to be consumed on a join.
+
+However, this does introduce a change in our security guarantees, and without
+additional tooling support and sensible defaults, the change may incentivize end
+users to create long-lived join tokens instead of using a more appropriate join
+method, or automating issuance of short lived tokens.
+
+To this end, we should introduce a per-bot-instance join count limit, and
+configure that to be 1 join by default. This matches today's behavior, and
+will help ensure users do not accidentally create a token that provides more
+access than expected: "infinite use" tokens with massive join limits and/or very
+long TTLs will need to be explicitly specified.
+
+We may additionally want to put hurdles in the way of things like extremely long
+token TTLs. There are legitimate low-security use cases for these, but we could
```

Comment:
I'm curious what Sasha thinks about this.

Extremely long join token TTLs basically become a shared secret like an API key, and we built Machine ID to avoid long-lived shared secrets..

Reply:
I agree, and would ideally like to only support this by way of an explicitly insecure `token-insecure` join method at some point in the future (see previous thread: https://github.com/gravitational/teleport/pull/36510#discussion_r1604165929). As we discussed in that thread, adding a TTL limit would _technically_ be a breaking change, and there are use cases we may still want to support.

For context on why, today's status quo is problematic because it seems like we properly support token joining, but it's kind of an illusion. It works great in a Getting Started guide on your dev box, and then breaks as soon as you try it in production because one of your nodes went offline for more than an hour. Production use really requires full automation (e.g. ansible or terraform detecting dead bots and reissuing tokens), and I'm not aware of any customers who've actually put in the effort to do it.

The changes proposed here (multiple uses per token, multiple tokens per bot) simplify the automation requirements a lot, but it's still a heavy lift, and I'm not sure it'll ever be worthwhile for a lot of use cases. There may be value in giving customers in this niche a way to still make use of Teleport if they (explicitly) accept the security implications of a long lived token; they still benefit from our RBAC, audit logging, etc.

Reply:
Michael is working on the display and creation of join tokens in the web UI, and we actually _just_ had a conversation with Sasha about token creation during the Design and Product Office Hours yesterday. In the meeting, we agreed to show a danger state for any `token` join method with an expiration >24 hours. We also agreed to make it impossible to create long-lived tokens through the web UI. 

(What we discussed doesn't affect the underlying mechanics of the `token` join method, but it does put some guard rails for their creation in the web UI and affects how they will be _displayed,_ regardless of how they were created.)